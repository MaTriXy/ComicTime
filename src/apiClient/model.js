export class NetworkError {}
export class UnknownError {
  constructor(statusCode) {
    this.statusCode = statusCode;
  }
}
export class NotFound {}
