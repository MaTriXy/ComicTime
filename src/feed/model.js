export class PublicationSummary {
  constructor(
    id,
    url,
    path,
    title,
    lastIssuesTitles,
    image,
    releaseDate,
    publisher
  ) {
    this.id = id;
    this.url = url;
    this.path = path;
    this.title = title;
    this.lastIssuesTitles = lastIssuesTitles;
    this.lastIssuesNumbers = lastIssuesTitles.map(issueTitle =>
      issueTitle.replace(title, "").trim()
    );
    this.image = image;
    this.releaseDate = releaseDate;
    this.publisher = publisher;
  }
}

export class Page {
  constructor(data, page, totalPages) {
    this.data = data;
    this.page = page;
    this.totalPages = totalPages;
  }
}
