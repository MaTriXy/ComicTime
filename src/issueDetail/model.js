export class IssuePage {
  constructor(number, image) {
    this.number = number;
    this.image = image;
  }
}
export class Issue {
  constructor(id, publicationId, title, numberOfPages, pages) {
    this.id = id;
    this.publicationId = publicationId;
    this.title = title;
    this.numberOfPages = numberOfPages;
    this.pages = pages;
  }
}
