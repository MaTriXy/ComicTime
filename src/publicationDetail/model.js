export class IssueSummary {
  constructor(id, title, releaseDate) {
    this.id = id;
    this.title = title;
    this.releaseDate = releaseDate;
  }
}
export class Publication {
  constructor(
    id,
    image,
    title,
    status,
    summary,
    publisher,
    releaseDate,
    authors,
    issues = []
  ) {
    this.id = id;
    this.image = image;
    this.title = title;
    this.status = status;
    this.summary = summary;
    this.publisher = publisher;
    this.releaseDate = releaseDate;
    this.authors = authors;
    this.issues = issues;
  }
}
