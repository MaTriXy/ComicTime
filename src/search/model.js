export class PublicationSearchResult {
  constructor(id, title) {
    this.id = id;
    this.title = title;
  }
}
