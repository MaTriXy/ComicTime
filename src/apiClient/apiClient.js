import rest from "rest";
import mime from "rest/interceptor/mime";
import log from "loglevel";
import { Left, Right } from "monet";
import { NetworkError, NotFound, UnknownError } from "./model";

const restClient = rest.wrap(mime);

export const baseUrl = "https://allorigins.me";

export function composeResourceUrl(resource) {
  return (
    "/get?url=" + encodeURIComponent(`http://readcomicsonline.ru${resource}`)
  );
}

export function get(resource) {
  return new Promise(resolve => {
    restClient(baseUrl + composeResourceUrl(resource))
      .then(response => {
        processResponse(response, resolve);
      })
      .catch(error => {
        processException(error, resolve);
      });
  });
}

function processResponse(response, resolve) {
  log.debug(response);
  const statusCode = response.status.code;
  if (statusCode === 200) {
    const pageHtml = response.entity.contents;
    resolve(Right(pageHtml));
  } else if (statusCode === 404) {
    resolve(Left(new NotFound()));
  } else {
    const unknownError = new UnknownError(statusCode);
    log.error(unknownError);
    resolve(Left(unknownError));
  }
}

function processException(error, resolve) {
  log.error(error);
  resolve(Left(new NetworkError()));
}
