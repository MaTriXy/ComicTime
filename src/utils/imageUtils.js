export async function downloadImageUrlAsBase64(url) {
  return new Promise((resolve, reject) => {
    let canvas = document.createElement("CANVAS");
    let img = document.createElement("img");
    img.setAttribute("crossorigin", "anonymous");
    img.src = "https://cors-anywhere.herokuapp.com/" + url;

    img.onload = () => {
      canvas.height = img.height;
      canvas.width = img.width;
      let context = canvas.getContext("2d");
      context.drawImage(img, 0, 0);
      let dataURL = canvas.toDataURL("image/jpeg", 0.8);
      canvas = null;
      resolve(dataURL);
    };
    img.onerror = () => {
      reject();
    };
  });
}
