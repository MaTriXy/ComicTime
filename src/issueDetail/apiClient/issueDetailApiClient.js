import { get } from "../../apiClient/apiClient";
import cheerio from "cheerio";
import { Issue, IssuePage } from "../model";
import JSZip from "jszip";
import { downloadImageUrlAsBase64 } from "../../utils/imageUtils";
import { formatNumberWithTwoDigits } from "../../utils/numberUtils";
import { downloadFile } from "../../utils/jszipUtils";
import { Right, Left } from "monet";
import log from "loglevel";

export async function getIssue(publicationId, issueId) {
  return get(`/comic/${publicationId}/${issueId}`).then(response => {
    return response.map(htmlPage =>
      extractIssueFromPage(publicationId, issueId, htmlPage)
    );
  });
}

export async function downloadIssueAsCbzFile(issue) {
  const fileName = `${issue.title}.cbz`;
  return generateZipFileFromIssueImages(issue)
    .then(eitherJszipFile => {
      return eitherJszipFile.map(jszipFile => {
        return downloadFile(jszipFile, fileName).then(file => Right(file));
      });
    })
    .catch(() => {
      return Left(fileName);
    });
}

function extractIssueFromPage(publicationId, issueId, htmlPage) {
  const loadedPage = cheerio.load(htmlPage);
  const title = loadedPage(".pager-cnt a")
    .first()
    .html();
  const numberOfPages = parseInt(
    loadedPage("#page-list").attr("data-size"),
    10
  );
  return new Issue(
    issueId,
    publicationId,
    title,
    numberOfPages,
    extractIssuePages(publicationId, issueId, numberOfPages)
  );
}

function extractIssuePages(publicationId, issueId, numberOfPages) {
  const issuePages = [];
  for (let page = 1; page <= numberOfPages; page++) {
    const completePageNumber = formatNumberWithTwoDigits(page);
    issuePages.push(
      new IssuePage(
        page,
        `http://readcomicsonline.ru/uploads/manga/${publicationId}/chapters/${issueId}/${completePageNumber}.jpg`
      )
    );
  }
  return issuePages;
}

async function generateZipFileFromIssueImages(issue) {
  const futureImages = issue.pages.map(page => {
    return downloadImageUrlAsBase64(page.image);
  });
  const imageNames = issue.pages.map(
    page => `${issue.title} - ${formatNumberWithTwoDigits(page.number)}.jpeg`
  );
  return Promise.all(futureImages)
    .then(base64Images => {
      return new Promise(resolve => {
        const zipFile = new JSZip();
        imageNames.forEach((name, index) => {
          const image = base64Images[index].split("base64,");
          zipFile.file(name, image[1], {
            base64: true,
            compression: "DEFLATE",
            compressionOptions: {
              level: 9
            }
          });
        });
        resolve(Right(zipFile));
      });
    })
    .catch(error => {
      log.error(`Error downloading issue: ${error}`);
      return Left();
    });
}
