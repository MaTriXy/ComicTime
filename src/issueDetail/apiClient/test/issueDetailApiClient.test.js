import {
  anyIssueId,
  anyPublicationId
} from "../../../testMothers/publicationDetails";

jest.mock("../../../utils/imageUtils", () => {
  return {
    downloadImageUrlAsBase64: jest.fn()
  };
});
jest.mock("../../../utils/jszipUtils", () => {
  return {
    downloadFile: jest.fn()
  };
});
import nock from "nock";
import { baseUrl, composeResourceUrl } from "../../../apiClient/apiClient";
import { NetworkError, UnknownError, NotFound } from "../../../apiClient/model";
import { downloadIssueAsCbzFile, getIssue } from "../issueDetailApiClient";
import { issueResponse } from "./resources/issueDetailApiClientResponses";
import { downloadImageUrlAsBase64 } from "../../../utils/imageUtils";
import { downloadFile } from "../../../utils/jszipUtils";
import { Right, Left } from "monet";
import { anyBase64Image } from "./resources/base64images";

describe("Issue detail API Client", () => {
  beforeEach(() => {
    downloadImageUrlAsBase64.mockReset();
    downloadFile.mockReset();
  });

  it("returns a not found error if the issue does not exist", async () => {
    const errorResponseCode = 404;
    givenTheIssueDoesNotExist(anyPublicationId, errorResponseCode);
    const result = await getIssue(anyPublicationId, anyIssueId);
    expect(result.left()).toEqual(new NotFound());
  });

  it("returns an error if there is something wrong", async () => {
    const errorResponseCode = 500;
    givenTheIssueDetailApiIsNotResponding(
      anyPublicationId,
      anyIssueId,
      errorResponseCode
    );
    const result = await getIssue(anyPublicationId, anyIssueId);
    expect(result.left()).toEqual(new UnknownError(errorResponseCode));
  });

  it("returns a network error if there is any not handled exception", async () => {
    givenThereIsNoConnection();
    const result = await getIssue(anyPublicationId, anyIssueId);
    expect(result.left()).toEqual(new NetworkError());
  });

  it("returns the extracted issue information composed by the list of pages", async () => {
    givenTheApiReturnsTheIssueInformation(anyPublicationId, anyIssueId);

    const result = await getIssue(anyPublicationId, anyIssueId);

    const issue = result.right();
    expect(issue.id).toEqual(anyPublicationId + "-" + anyIssueId);
    expect(issue.title).toEqual("The Flash (2016-) #Annual 1");
    expect(issue.numberOfPages).toEqual(43);
    const firstPage = issue.pages[0];
    expect(firstPage.number).toEqual(1);
    expect(firstPage.image).toEqual(
      "https://readcomicsonline.ru/uploads/manga/the-flash-2016/chapters/Annual1/01.jpg"
    );
  });

  it("returns an issue with all the pages found", async () => {
    givenTheApiReturnsTheIssueInformation(anyPublicationId, anyIssueId);

    const result = await getIssue(anyPublicationId, anyIssueId);
    expect(result.right()).toMatchSnapshot();
  });

  it("generates a zip file using the pages of an issue already loaded", async () => {
    givenTheApiReturnsTheIssueInformation(anyPublicationId, anyIssueId);
    const getIssueResult = await getIssue(anyPublicationId, anyIssueId);
    const issue = getIssueResult.right();
    const fileName = `${issue.title}.cbz`;
    givenThatIssuePagesAreDownloadedProperly(fileName);

    const file = await downloadIssueAsCbzFile(issue, jest.fn());

    expect(file.isRight()).toBeTruthy();
  });

  it("returns an error if any of the images being downloaded fail", async () => {
    givenTheApiReturnsTheIssueInformation(anyPublicationId, anyIssueId);
    const getIssueResult = await getIssue(anyPublicationId, anyIssueId);
    const issue = getIssueResult.right();
    givenTheIssuePagesAreNotDownloadedProperly();

    const file = await downloadIssueAsCbzFile(issue);

    expect(file.isRight()).toBeFalsy();
  });
});

function givenTheIssuePagesAreNotDownloadedProperly() {
  downloadFile.mockReturnValue(Promise.resolve(Left()));
}

function givenThatIssuePagesAreDownloadedProperly(fileName) {
  const result = Promise.resolve(anyBase64Image);
  downloadImageUrlAsBase64.mockReturnValue(result);
  downloadFile.mockReturnValue(Promise.resolve(Right(fileName)));
}

function givenTheIssueDoesNotExist(publicationId, issueId) {
  nock(baseUrl)
    .get(composeResourceUrl(`/comic/${publicationId}/${issueId}`))
    .reply(404);
}

function givenTheIssueDetailApiIsNotResponding(
  publicationId,
  issueId,
  errorStatusCode
) {
  nock(baseUrl)
    .get(composeResourceUrl(`/comic/${publicationId}/${issueId}`))
    .reply(errorStatusCode);
}

function givenThereIsNoConnection() {
  nock(baseUrl)
    .get("/notHandledPath")
    .reply(1);
}

function givenTheApiReturnsTheIssueInformation(publicationId, issueId) {
  nock(baseUrl)
    .get(composeResourceUrl(`/comic/${publicationId}/${issueId}`))
    .reply(200, issueResponse);
}
