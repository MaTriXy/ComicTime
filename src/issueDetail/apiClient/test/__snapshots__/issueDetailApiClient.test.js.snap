// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Issue detail API Client returns an issue with all the pages found 1`] = `
Issue {
  "id": "the-flash-2016-Annual1",
  "numberOfPages": 20,
  "pages": Array [
    IssuePage {
      "image": "https://readcomicsonline.ru/uploads/manga/the-flash-2016/chapters/Annual1/01.jpg",
      "number": 1,
    },
    IssuePage {
      "image": "https://readcomicsonline.ru/uploads/manga/the-flash-2016/chapters/Annual1/02.jpg",
      "number": 2,
    },
    IssuePage {
      "image": "https://readcomicsonline.ru/uploads/manga/the-flash-2016/chapters/Annual1/03.jpg",
      "number": 3,
    },
    IssuePage {
      "image": "https://readcomicsonline.ru/uploads/manga/the-flash-2016/chapters/Annual1/04.jpg",
      "number": 4,
    },
    IssuePage {
      "image": "https://readcomicsonline.ru/uploads/manga/the-flash-2016/chapters/Annual1/05.jpg",
      "number": 5,
    },
    IssuePage {
      "image": "https://readcomicsonline.ru/uploads/manga/the-flash-2016/chapters/Annual1/06.jpg",
      "number": 6,
    },
    IssuePage {
      "image": "https://readcomicsonline.ru/uploads/manga/the-flash-2016/chapters/Annual1/07.jpg",
      "number": 7,
    },
    IssuePage {
      "image": "https://readcomicsonline.ru/uploads/manga/the-flash-2016/chapters/Annual1/08.jpg",
      "number": 8,
    },
    IssuePage {
      "image": "https://readcomicsonline.ru/uploads/manga/the-flash-2016/chapters/Annual1/09.jpg",
      "number": 9,
    },
    IssuePage {
      "image": "https://readcomicsonline.ru/uploads/manga/the-flash-2016/chapters/Annual1/10.jpg",
      "number": 10,
    },
    IssuePage {
      "image": "https://readcomicsonline.ru/uploads/manga/the-flash-2016/chapters/Annual1/11.jpg",
      "number": 11,
    },
    IssuePage {
      "image": "https://readcomicsonline.ru/uploads/manga/the-flash-2016/chapters/Annual1/12.jpg",
      "number": 12,
    },
    IssuePage {
      "image": "https://readcomicsonline.ru/uploads/manga/the-flash-2016/chapters/Annual1/13.jpg",
      "number": 13,
    },
    IssuePage {
      "image": "https://readcomicsonline.ru/uploads/manga/the-flash-2016/chapters/Annual1/14.jpg",
      "number": 14,
    },
    IssuePage {
      "image": "https://readcomicsonline.ru/uploads/manga/the-flash-2016/chapters/Annual1/15.jpg",
      "number": 15,
    },
    IssuePage {
      "image": "https://readcomicsonline.ru/uploads/manga/the-flash-2016/chapters/Annual1/16.jpg",
      "number": 16,
    },
    IssuePage {
      "image": "https://readcomicsonline.ru/uploads/manga/the-flash-2016/chapters/Annual1/17.jpg",
      "number": 17,
    },
    IssuePage {
      "image": "https://readcomicsonline.ru/uploads/manga/the-flash-2016/chapters/Annual1/18.jpg",
      "number": 18,
    },
    IssuePage {
      "image": "https://readcomicsonline.ru/uploads/manga/the-flash-2016/chapters/Annual1/19.jpg",
      "number": 19,
    },
    IssuePage {
      "image": "https://readcomicsonline.ru/uploads/manga/the-flash-2016/chapters/Annual1/20.jpg",
      "number": 20,
    },
  ],
  "publicationId": "the-flash-2016",
  "title": "The Flash (2016-) #Annual 1",
}
`;
